### PR TITLE
Mirror before scaling in _AnimatedIconPainter

### LINKED ISSUE #60521: Several AnimatedIcons are positioned incorrectly in RTL
## Steps to Reproduce

1. Create an app that uses one of:
    * `AnimatedIcons.menu_arrow`
    * `AnimatedIcons.arrow_menu`
    * `AnimatedIcons.list_view`
    * `AnimatedIcons.view_list`
2. Set Directionality to RTL
3. Observe incorrect positioning of Icon (see attached screenshot)

Reproducible example here: https://dartpad.dev/87b00d935692791575685eb34074f2eb

**Expected results:** 
Icon is positioned in the center of its parent as expected

**Actual results:** 
Icon is shifted to the top left (see screenshot)

Simulator Screen Shot - iPhone 11 Pro - 2020-06-29 at 10 24 29

### PR DESCRIPTION
**Pull request by Amir-P**
This PR attempts to fix an issue in `_AnimatedIconPainter` causing wrong positioning of `_AnimatedIconData` when `matchTextDirection` set true in layouts with RTL directionality. This can be fixed by doing mirror before scaling the canvas.

### Code to reproduce:
```dart
void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Directionality(
        textDirection: TextDirection.rtl,
        child: Scaffold(
          appBar: AppBar(
            leading: IconButton(
              icon: AnimatedIcon(
                icon: AnimatedIcons.arrow_menu,
                progress: AlwaysStoppedAnimation(0),
              ),
              onPressed: () {},
            ),
          ),
        ),
      ),
    );
  }
}
```

Before:

After:

### Issues this PR attempts to fix:
Fixes #60521 

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Review by Piinks — COMMENTED** [review]
Hi @Amir-P thanks for the contribution, and welcome! 
Can you add another test that validates the positioning like from the example before/after images you shared?
Also, it looks like there are currently failing tests, can you take a look?

**Comment by Amir-P** [comment]
> Hi @Amir-P thanks for the contribution, and welcome! 
> 
> Can you add another test that validates the positioning like from the example before/after images you shared?
> 
> Also, it looks like there are currently failing tests, can you take a look?

Thanks! @Piinks 
I've fixed the issue with private tests. One of the checks (tree check) is still failing and I think that's not because of my changes.
Also I couldn't find a way to test the final position of drawn path. The position is determined by the transformations made to the canvas before drawing path and we have tests for that.
If you can guide me on writing tests for the position of path on canvas, please let me know.

**Review by Piinks — CHANGES_REQUESTED** [review]
> Also I couldn't find a way to test the final position of drawn path. The position is determined by the transformations made to the canvas before drawing path and we have tests for that.
If you can guide me on writing tests for the position of path on canvas, please let me know.

The type of test to add should take the code sample that illustrates your bug (above), and verify that the icon is placed in the right position now. This validates that it actually fixes the reported bug, and prevents it from regressing in the future. 
Here is a similar test: https://github.com/flutter/flutter/blob/0e2cc64f2e545474993212d02fc08fddc8dc739f/packages/flutter/test/material/app_bar_test.dart#L566

Instead of checking the size (`tester.getSize`), you would want to check the position, using `tester.getCenter` or `tester.getTopLeft` and validate that the icon is in the right place now.

**Comment by Amir-P** [comment]
> Instead of checking the size (`tester.getSize`), you would want to check the position, using `tester.getCenter` or `tester.getTopLeft` and validate that the icon is in the right place now.

This change does not effect the widget and it's position. You can check the `AnimatedIcon` widget's position and size in both RTL and LTR direction and see there is no difference. It's happening in the painting level on canvas. You can validate the behavior using this small test (Before and after my commit):
```dart
testWidgets('Direction has no effect on AnimatedIcon widget position',
      (WidgetTester tester) async {
    await tester.pumpWidget(
      const Directionality(
        textDirection: TextDirection.rtl,
        child: AnimatedIcon(
          progress: AlwaysStoppedAnimation<double>(0.0),
          icon: AnimatedIcons.arrow_menu,
        ),
      ),
    );
    final Rect rtlRect = tester.getRect(find.byType(AnimatedIcon));
    await tester.pumpWidget(
      const Directionality(
        textDirection: TextDirection.ltr,
        child: AnimatedIcon(
          progress: AlwaysStoppedAnimation<double>(0.0),
          icon: AnimatedIcons.arrow_menu,
        ),
      ),
    );
    final Rect ltrRect = tester.getRect(find.byType(AnimatedIcon));
    expect(rtlRect, ltrRect);
  });
```
@Piinks

**Review by Piinks — COMMENTED** [review]
> ```dart
> testWidgets('Direction has no effect on AnimatedIcon widget position',
>       (WidgetTester tester) async {
>     await tester.pumpWidget(
>       const Directionality(
>         textDirection: TextDirection.rtl,
>         child: AnimatedIcon(
>           progress: AlwaysStoppedAnimation<double>(0.0),
>           icon: AnimatedIcons.arrow_menu,
>         ),
>       ),
>     );
>     final Rect rtlRect = tester.getRect(find.byType(AnimatedIcon));
>     await tester.pumpWidget(
>       const Directionality(
>         textDirection: TextDirection.ltr,
>         child: AnimatedIcon(
>           progress: AlwaysStoppedAnimation<double>(0.0),
>           icon: AnimatedIcons.arrow_menu,
>         ),
>       ),
>     );
>     final Rect ltrRect = tester.getRect(find.byType(AnimatedIcon));
>     expect(rtlRect, ltrRect);
>   });
> ```

I couldn't find this test in the framework, can you add it? Thanks!

Screen Shot 2022-01-04 at 1 16 16 PM

Also, we'll still need a test that validates the very real use case you have in your before and after image above. Otherwise we may regress in the future. The current test validates that you have changed the order of operations, not that the visual difference above has changed.

If checking the position does not validate this, then it sounds like a golden file test is more appropriate. Can you add one? Thanks!

**Comment by Amir-P** [comment]
I've found this [document](https://github.com/flutter/flutter/wiki/Writing-a-golden-file-test-for-package:flutter) on writing golden file tests but It doesn't say how can I create and add a new image for this case. Can you help me on this? @Piinks

**Comment by Amir-P** [comment]
Can you please validate the latest commit? Some of the tests are failing but I'm not sure why because I didn't make any changes. Also I guess we need to upload those golden file images which I don't have access to. @Piinks

**Comment by skia-gold** [comment]
Gold has detected about 6 new digest(s) on patchset 3.
View them at https://flutter-gold.skia.org/cl/github/93312

File: packages/flutter/test/material/animated_icons_test.dart [hunk 1]
```diff
@@ -252,10 +271,15 @@ void main() {
...
+    expect(canvas.invocations, const <RecordedCanvasCall>[
+      RecordedScale(0.5, 0.5),
+    ]);
+    await expectLater(find.byType(AnimatedIcon),
+        matchesGoldenFile('animated_icons_test.icon.ltr.png'));
   });
 
-  testWidgets('Inherited text direction overridden', (WidgetTester tester) async {
+  testWidgets('Inherited text direction overridden',
+      (WidgetTester tester) async {
```
Comment (Piinks) [inline packages/flutter/test/material/animated_icons_test.dart:282]: Nit: Can you revert this style change?

File: packages/flutter/test/material/animated_icons_test.dart [hunk 2]
```diff
@@ -277,8 +301,26 @@ void main() {
     expect(canvas.invocations, const <RecordedCanvasCall>[
       RecordedRotate(math.pi),
       RecordedTranslate(-48, -48),
+      RecordedScale(0.5, 0.5),
     ]);
   });
+
+  testWidgets('Direction has no effect on position of widget',
+      (WidgetTester tester) async {
```
Comment (Piinks) [inline packages/flutter/test/material/animated_icons_test.dart:309]: Nit: This should be one line

File: packages/flutter/test/material/animated_icons_test.dart [hunk 3]
```diff
@@ -292,7 +334,7 @@ class PaintColorMatcher extends Matcher {
 
   @override
   Description describe(Description description) =>
-    description.add('color was not $expectedColor');
+      description.add('color was not $expectedColor');
```
Comment (Piinks) [inline packages/flutter/test/material/animated_icons_test.dart:337]: Nit: revert this, these style changes don't match the style guide.

**Review by Piinks — COMMENTED** [review]
If you visit https://flutter-gold.skia.org/cl/github/93312 you can see the images your test generated for each platform, I can approve them when this is ready to land. You may want to add a RepaintBoundary around the icon widget, right now the images are large white squares with a tiny icon in the top left corner. :)

It looks like the file with the golden file test needs a tag added at the top to pass analysis, the other failing checks should be fixed by updating your branch.

For more on golden file testing, check out this wiki page:
https://github.com/flutter/flutter/wiki/Writing-a-golden-file-test-for-package:flutter

File: packages/flutter/test/material/animated_icons_test.dart (hunk 2 above)
Comment (Amir-P) [reply packages/flutter/test/material/animated_icons_test.dart:309]: You mean to remove the break line after `...widget',` ?

Comment (Piinks) [reply packages/flutter/test/material/animated_icons_test.dart:309]: Yup!

**Comment by Amir-P** [comment]
> If you visit https://flutter-gold.skia.org/cl/github/93312 you can see the images your test generated for each platform, I can approve them when this is ready to land. You may want to add a RepaintBoundary around the icon widget, right now the images are large white squares with a tiny icon in the top left corner. :)
> 
> It looks like the file with the golden file test needs a tag added at the top to pass analysis, the other failing checks should be fixed by updating your branch.
> 
> For more on golden file testing, check out this wiki page: https://github.com/flutter/flutter/wiki/Writing-a-golden-file-test-for-package:flutter

Hope I got it right this time🤞🏻. @Piinks

**Comment by skia-gold** [comment]
Gold has detected about 7 new digest(s) on patchset 4.
View them at https://flutter-gold.skia.org/cl/github/93312

**Comment by Piinks** [comment]
Coming along nicely! Thanks for the updates! It looks like you still need to update your PR with the latest from the master branch. :)

**Comment by skia-gold** [comment]
Gold has detected about 4 new digest(s) on patchset 4.
View them at https://flutter-gold.skia.org/cl/github/93312

**Comment by Amir-P** [comment]
> Coming along nicely! Thanks for the updates! It looks like you still need to update your PR with the latest from the master branch. :)

Done. @Piinks

**Review by Piinks — APPROVED** [review]
Golden file images are approved, and this LGTM! Thank you!

**Comment by Piinks** [comment]
This will need a secondary review, I've reached out to the team for one. Meanwhile, can you take a look at the merge conflict here?

**Comment by skia-gold** [comment]
Gold has detected about 4 new digest(s) on patchset 5.
View them at https://flutter-gold.skia.org/cl/github/93312

**Comment by Amir-P** [comment]
Done @Piinks .

File: packages/flutter/test/material/animated_icons_test.dart [hunk 4]
```diff
@@ -2,6 +2,8 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 
+@Tags(<String>['reduced-test-set'])
```
Comment (Hixie) [inline packages/flutter/test/material/animated_icons_test.dart:5]: what does this do?

**Review by Hixie — APPROVED** [review]
LGTM

File: packages/flutter/test/material/animated_icons_test.dart (hunk 4 above)
Comment (Piinks) [reply packages/flutter/test/material/animated_icons_test.dart:5]: flutter.dev/go/reduce-ci-tests
https://github.com/flutter/flutter/wiki/Writing-a-golden-file-test-for-package:flutter#reduced-test-set-tag

One of the velocity projects over the summer involved removing unit test runs on windows and mac machines. Just doing so though would remove our ability to provide a golden file test for every platform someone might be contributing to Flutter on, so we created a reduced test set for windows and mac. The analyzer will ensure now that all test files that contain a golden file test are included in the reduced test set.

Comment (Hixie) [reply packages/flutter/test/material/animated_icons_test.dart:5]: oh nice

**Pull request merged**